File: 0138.CopyListwithRandomPointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if not head:
            return
        p = head
        while p:
            pp = RandomListNode(p.label)
            pp.next = p.next
            p.next = pp
            p = pp.next

        p = head
        while p:
            pp = p.next
            if p.random:
                pp.random = p.random.next
            p = pp.next

        headNew = head.next
        p, pp = head, headNew
        while pp.next:
            p.next = pp.next
            p = p.next
            pp.next = p.next
            pp = pp.next
        p.next = None
        #pp.next = None

        return headNew

File: 0000-0999/0138.CopyListwithRandomPointer.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Next *Node
 *     Random *Node
 * }
 */

func copyRandomList(head *Node) *Node {
    if head == nil {
        return nil
    }
    p := head
    for p != nil {
        q := &Node{Val: p.Val, Next: p.Next}
        p.Next = q
        p = q.Next
    }
    p = head
    for p != nil {
        q := p.Next
        if p.Random != nil {
            q.Random = p.Random.Next
        }
        p = q.Next
    }
    newHead := head.Next
    p, q := head, newHead
    for q.Next != nil {
        p.Next = q.Next
        p = p.Next
        q.Next = p.Next
        q = q.Next
    }
    p.Next = nil
    return newHead
}